smoketest/alertlog_test.go

```diff
@@ -0,0 +1,252 @@
+	t.Parallel()
+
+	var sql = makeSQL(true, "SMS", true, true, true)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var details = logs.Alert.RecentEvents.Nodes[0].State.Details
+	assert.Equal(t, "contact method disabled", details)
+}
+
+func TestSMSFailure(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "SMS", true, true, true)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Twilio(t).Device(h.Phone("1")).RejectSMS("Alert #1: foo")
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var msg = logs.Alert.RecentEvents.Nodes[0].Message
+	var details = logs.Alert.RecentEvents.Nodes[0].State.Details
+	assert.Contains(t, msg, "Notification sent")
+	assert.Equal(t, "failed", details)
+}
+
+func TestVoiceFailure(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "VOICE", true, true, true)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Twilio(t).Device(h.Phone("1")).RejectVoice("foo")
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var msg = logs.Alert.RecentEvents.Nodes[0].Message
+	var details = logs.Alert.RecentEvents.Nodes[0].State.Details
+	assert.Contains(t, msg, "Notification sent")
+	assert.Equal(t, "failed", details)
+}
+
+func TestNoImmediateNR(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "SMS", false, true, true)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var msg = logs.Alert.RecentEvents.Nodes[0].Message
+	assert.Contains(t, msg, "no immediate rule")
+}
+
+func TestNoOnCallUsers(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "SMS", true, true, false)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var details = logs.Alert.RecentEvents.Nodes[0].State.Details
+	assert.Equal(t, "No one was on-call", details)
+}
+
+func TestNoEPSteps(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "SMS", true, false, false)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var details = logs.Alert.RecentEvents.Nodes[0].State.Details
+	assert.Equal(t, "No escalation policy steps", details)
+}
+
+func makeSQL(disabledCM bool, cmType string, nr bool, epStep bool, epStepUser bool) string {
```

Comment:
Let's make this a struct so it's easier to read where it's used.

```go
type alertLogTestCfg struct {
  CMDisabled bool
  CMType string
  NR bool
  EPStep bool
  EPStepUser bool
}
```

Lastly, what do you think of allowing a data parameter for the initSQL?

```go
const alertLogSQLTmpl = `
		insert into users (id, name, email) 
		values ({{uuid "user"}}, 'bob', 'joe');
		insert into user_contact_methods (id, user_id, name, type, value, disabled) 
		values ({{uuid "cm1"}}, {{uuid "user"}}, 'personal', '{{.CMType}}', {{phone "1"}}, {{.CMDisabled}});

		{{- if .NR}}
		insert into user_notification_rules (user_id, contact_method_id, delay_minutes) 
		values ({{uuid "user"}}, {{uuid "cm1"}}, 0);
		{{- end}}
`
h := harness.NewHarnessWithData(t, alertLogSQLTmpl, "add-no-notification-alert-log", alertLogTestCfg{
  CMType: "SMS",
  NR: true,
  EPStep: true,
  EPStepUser: true,
})
```


I think it would be a lot cleaner to have the if logic as part of the template. The `Harness.execQuery` method can just add a `data interface{}` argument and pass that to `Execute` instead of the hard-coded `nil` value and it should be good.